### PR TITLE
Added skeleton x-ray

### PR DESCRIPTION
**Pull request by AndreaCatania**
I've added the x-ray on skeleton in order to see its bones. It can be activated in the Debug menu.

untitled2

I'm not 100% sure that EditorNode is the right way where to put the variable debug_skeleton_xrays. Please let me know if I've to move it in a better place.

My intent was to add the skeleton bones names, but it require the creation of an additional Node that allow me to use a label in the 3D world. So for now I'm submitting only this.

Related topic: https://github.com/godotengine/godot/issues/11507

### CODE REVIEW

**Comment by n-pigeon** [comment]
Will this make bones visible in game? This menu has options to display debug stuff when game runs.

I would change name a bit - "Visible Skeleton Bones" ?

**Comment by AndreaCatania** [comment]
Yes you true, The name does not explain what it does very well, instead "Visible Skeleton Bones" sounds very good to me. 

Also this material is set in the Skeleton gizmo and it's loaded only into the editor so the bones are never rendered in game.
Initially I was implementing it in the skeleton menu (that doesn't exist in this moment) but I thought that it's way more handy had it in a menu that is always visible, So I move it there. Probably the best position could be under the _Editor_ Menu.

**Comment by n-pigeon** [comment]
If it's like that I think it should be in the View menu of the Spatial Editor. It is probably the most logical place for 3D viewport feature like this.

**Comment by Noshyaar** [comment]
The current location of the option implies that it will be visible during runtime. (which imo a nice feature, but I'm not sure about actual use case.) If that's not the case, I agree that SpatialEditor toolbar or its submenus are better places.

**Comment by Zylann** [comment]
It could be added to Skeleton node maybe? This way we would be able to see bones when debugging a game too (like collision shapes), just suggesting

**Comment by groud** [comment]
Ok some feedback from the PR monday:
- having as an entry of the debug menu is probably not a good choice.
- @reduz suggested we could instead have an option to make all gizmos x-ray, thus it could be a viewport option instead :).

**Comment by AndreaCatania** [comment]
I'll put it in the viewport menu, as suggested by n-pigeon. I didn't have the time to do that before, I'll update it soon. I'll use "Gizmo x-ray" if the reduz's idea is to have x-ray for all gizmos in the editor.

**Comment by AndreaCatania** [comment]
I've moved the option in the View menu.
When you click Gizmos X-Ray is required that you click the skeleton node in the nodes panel in order to redraw the gizmos with new material. Do you know how can I force Gizmo to be redrawed?

**Comment by Hinsbart** [comment]
From todays PR irc meeting:
<reduz> this should probably be applied to all gizmos, not skeleton

**Comment by reduz** [comment]
Spatial objects have an update_gizmo() function to update the gizmos

**Comment by groud** [comment]
Also, what's debug_skeleton_bones_names for ? it doesn't seem used.

**Comment by AndreaCatania** [comment]
I've changed menu name to "Show bones on top"
Removed unused code
Implemented function that update gizmos to instantly show or hide bones.

**Comment by AndreaCatania** [comment]
Do you know when this can be merged? I need it

**Comment by groud** [comment]
Well there are still some problem with the PR. Please,
- Make sure the change affects all 3D gizmos.
- Remove `set_skeleton_bones_names()`, `get_debug_skeleton_bones_names()` and `debug_skeleton_bones_names` as they are not used.
- Rename `set_gizmos_xrays()` to `set_show_gizmos_on_top()`,` is_debug_skeleton_xrays()` to `is_show_gizmos_on_top()` and `debug_skeleton_xrays` to `show_gizmos_on_top`
- Remove the `update_all_gizmos()` function, and just call `_update_all_gizmos()` at the end of `set_show_gizmos_on_top()`.

**Comment by AndreaCatania** [comment]
I was checking this PR code, and I saw that it went in a different direction of my first intent.

My main need was to view skeleton bones (x-ray) and at the same time I want to see only skeleton bones on top in order to avoid confusion in the editor.

untitled

For example the image above show only the skeleton on top but the scene is to crowded, imagine if all gizmos were on top, it'd be chaos.

@Hinsbart said:
> From todays PR irc meeting:
> [reduz] this should probably be applied to all gizmos, not skeleton

Now I want revert it and focus just to the original thing, so @reduz what do you think about the implementation of an option for each spatial gizmo that require to stay on top?

In order to have 
- Show skeleton on top
- Show XXXXX on top
- Show YYYY on top

and allow the user to customize the view as he needs.

**Comment by groud** [comment]
Ok, if it bloat the view it's not a good idea then.
In fact I think having this as an option is a little bit unexpected. Is there any reason why we would not want the bone to be drawn on the top ?

**Comment by AndreaCatania** [comment]
@groud always on top? I want to add more control to the view.

My idea is to add some other options in order to show / hide gizmos.

For example in the picture above would been usefull hide all collisions gizmos, sice I was working on skeleton.

So put something always on top is not so usefull.

**Comment by groud** [comment]
@AndreaCatania, maybe not always on top, but when you select a node from the same skeleton (maybe ?).

My point is that we should avoid adding "optional" stuff if one of the option is never used. If it's hard working with bones which are not drawn on top, just make "drawn on top" the default, then maybe add an option to show/hide the bones totally.

**Comment by AndreaCatania** [comment]
> My point is that we should avoid adding "optional" stuff if one of the option is never used.

You can't assume that one option is optional and will never be used only because it doesn't yet exist, or you don't need it. If I did this PR is because I need it and this is a prove that it is not useless.

I don't exactly know why you are against it but probably you are scared about adding too many options, that for sure give more control to the user.

This screenshot is how looks like UE4, it has tons of options that give all kind of required customizations (It is a completed editor, so we can base to it).
untitled

Another thing that I want to point out is that, completely hide a gizmo or show it always on top is could be very annoying, I think we have to give more control to user.

**Comment by groud** [comment]
To be honest, I think Unreal is doing a pretty bad job about this... Their interface is overcrowded with a lot of useless elements that are everything but user-friendly. Having too much options is often useless (as nobody change 90% of them) or at least confusing as it's hard to find the truly interesting ones. As for the API, we should be careful about what we let available to the user and what not, as we want Godot to stay easy to learn.

I am not against the option, but if there is a way to avoid something to be optional I think it is better. I have the feeling that bones are pretty useless if not drawn on top, as if they are hidden inside an object (which may happen most of the time) it is like is there were not here. Thus, if we can find a way to hide them when they are not needed, I think it is better. A solution could be to hide the bones when the object is not selected (but I'm open to other proposals :) ).

I'm not against merging, but for me this option does not seem very ergonomic as is. I think it need more opinions.

> If I did this PR is because I need it and this is a prove that it is not useless.

Be careful, you're more than the everyday user. Debugging the editor/engine is not like creating a game with the engine. :)

**Comment by reduz** [comment]
Instead of adding options, why dont we simply make the skeleton be always
on top? I dont see any use for it being behind objects. In fact, I think
this is how it used to work before..

**Comment by AndreaCatania** [comment]
@groud UE4 is too crowded and this is not a good thing when you open it for the first time, and here I agree with you. And yes my formamentis is different from every day user and my opinion could be wrong.

However depending on the situation it may be useful have it visible, on top, or hidden. Probably a button with three way check could be useful.

Something like: 
untitled

In this way we can save options numbers and at the same time we get full control of all gizmos, if you don't think it's a good idea, draw the skeleton always on top as default and allow to hide it with an option is enought for my needs.

**Comment by n-pigeon** [comment]
@AndreaCatania I think this is very good idea.

**Comment by groud** [comment]
I'm ok with such option for each gizmo, still I don't know see any use case where you need to have the bones visible but not drawn on the top.

**Comment by AndreaCatania** [comment]
It's a rare case, but for example if you need to set the size of physical bone, or set the transform of attached node could be hard in some cases if the skeleton is drawd on top.

Imagine if you want to do it for an hand where you have 20 bones drawed all togheter.

**Comment by n-pigeon** [comment]
We can always make it so bones are always visible on top and check if that really is a problem or not.
I think at minimum, we have to be able to hide them without turning them OFF.

Although if we make clean feature that some nodes if they need to can be in XRay mode using Scene tab, then it will be nice addition for usability. I see this useful not only for Bones.

**Comment by groud** [comment]
> It's a rare case, but for example if you need to set the size of physical bone, or set the transform of attached node could be hard in some cases if the skeleton is drawn on top.

Yeah but in this case, just hiding them would do the trick, wouldn't it ? ^^

> I think at minimum, we have to be able to hide them without turning them OFF.

I agree on that :)

In any case, my point is that every option should have a purpose corresponding to a specific use case.  For example, I'd rather have a "Debug collisions" checkbox, that displays collision boxes on the top, displays a different color when collisions occur or displays debug text, rather than several checkboxes to activate the features one by one.

**Comment by AndreaCatania** [comment]
I've implemented it. This is how looks like:

skeleton_x-ray

I like it very much.

File: scene/gui/popup_menu.cpp [hunk 1]
```diff
@@ -1098,8 +1161,10 @@ void PopupMenu::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("set_item_as_checkable", "idx", "enable"), &PopupMenu::set_item_as_checkable);
 	ClassDB::bind_method(D_METHOD("set_item_tooltip", "idx", "tooltip"), &PopupMenu::set_item_tooltip);
 	ClassDB::bind_method(D_METHOD("set_item_shortcut", "idx", "shortcut", "global"), &PopupMenu::set_item_shortcut, DEFVAL(false));
+	ClassDB::bind_method(D_METHOD("set_item_statable", "idx", "state"), &PopupMenu::set_item_statable);
```
Comment (akien-mga) [inline scene/gui/popup_menu.cpp]: I think `multistate` might be easier to understand than `statable`. I'll merge anyway but we could improve that in another PR.

**Pull request merged**